File: saddle-points/src/saddle_points.erl
-module(saddle_points).

-export([saddle_points/1, index_max_per_row/1]).

rows(Matrix) ->
    Matrix.

transpose([[] | _]) ->
    [];
transpose(M) ->
    [lists:map(fun hd/1, M) | transpose(lists:map(fun tl/1, M))].

columns(Matrix) ->
    transpose(Matrix).

with_index(Matrix) ->
    lists:zip(
        lists:seq(0, length(Matrix) - 1), Matrix).

index_max_per_row(Rows) ->
    sets:from_list(
        lists:flatmap(fun({RIndex, Row}) ->
                         Max = lists:max(Row),
                         lists:map(fun({CIndex, _Value}) -> {RIndex, CIndex} end,
                                   lists:filter(fun({_CIndex, Value}) -> Value == Max end,
                                                with_index(Row)))
                      end,
                      with_index(Rows))).

index_min_per_columns(Columns) ->
    sets:from_list(
        lists:flatmap(fun({CIndex, Column}) ->
                         Min = lists:min(Column),
                         lists:map(fun({RIndex, _Value}) -> {RIndex, CIndex} end,
                                   lists:filter(fun({_RIndex, Value}) -> Value == Min end,
                                                with_index(Column)))
                      end,
                      with_index(Columns))).

saddle_points([[]]) ->
    [];
saddle_points(Matrix) ->
    Rows = rows(Matrix),
    Columns = columns(Matrix),
    sets:to_list(
        sets:intersection(index_max_per_row(Rows), index_min_per_columns(Columns))).

File: elixir/saddle-points/lib/saddle_points.ex
defmodule SaddlePoints do
  @doc """
  Parses a string representation of a matrix
  to a list of rows
  """
  @spec rows(String.t()) :: [[integer]]
  def rows(str) do
    str
    |> String.split("\n")
    |> Enum.map(fn line -> line |> String.split() |> Enum.map(&String.to_integer/1) end)
  end

  defp transpose(rows) do
    rows
    |> List.zip()
    |> Enum.map(&Tuple.to_list/1)
  end

  @doc """
  Parses a string representation of a matrix
  to a list of columns
  """
  @spec columns(String.t()) :: [[integer]]
  def columns(str) do
    rows(str) |> transpose()
  end

  @doc """
  Calculates all the saddle points from a string
  representation of a matrix
  """
  @spec saddle_points(String.t()) :: [{integer, integer}]
  def saddle_points(str) do
    rows = rows(str)
    columns = columns(str)

    indexes_max_per_row =
      rows
      |> Stream.with_index()
      |> Enum.flat_map(fn {row, r_i} ->
        max = row |> Enum.max()

        row
        |> Stream.with_index()
        |> Stream.filter(fn
          {^max, _index} -> true
          _ -> false
        end)
        |> Enum.map(fn {_v, c_i} -> {r_i, c_i} end)
      end)
      |> MapSet.new()

    indexes_min_per_column =
      columns
      |> Stream.with_index()
      |> Enum.flat_map(fn {column, c_i} ->
        min = column |> Enum.min()

        column
        |> Stream.with_index()
        |> Stream.filter(fn
          {^min, _index} -> true
          _ -> false
        end)
        |> Enum.map(fn {_v, r_i} -> {r_i, c_i} end)
      end)
      |> MapSet.new()

    MapSet.intersection(indexes_max_per_row, indexes_min_per_column) |> MapSet.to_list()
  end
end
